File: java/src/0013.java
import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

class Solution {

    public int romanToInt(String s) {
        var n = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            n += switch (s.charAt(i)) {
                case 'I' -> n < 5 ? 1 : -1;
                case 'V' -> 5;
                case 'X' -> n < 50 ? 10 : -10;
                case 'L' -> 50;
                case 'C' -> n < 500 ? 100 : -100;
                case 'D' -> 500;
                default -> 1000;
            };
        }
        return n;
    }

    public static void main(String[] args) {
        record Case(String s, int excepted) {
        }

        var CASES = new Case[] {
                new Case("III", 3),
                new Case("LVIII", 58),
                new Case("MCMXCIV", 1994),
        };

        var o = new Solution();

        for (var x : CASES) {
            assertThat(o.romanToInt(x.s)).isEqualTo(x.excepted);
        }

        var start = Instant.now();
        for (var x : CASES) {
            for (var i = 0; i < 1000000; i++) {
                o.romanToInt(x.s);
            }
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}

File: python/src/0013.py
# yapf: disable
class Solution:
    def romanToInt(self, s: str) -> int:
        n = 0
        p = ''
        for c in s:
            match c:
                case 'M': n += 800 if p == 'C' else 1000
                case 'D': n += 300 if p == 'C' else 500
                case 'C': n += 80 if p == 'X' else 100
                case 'L': n += 30 if p == 'X' else 50
                case 'X': n += 8 if p == 'I' else 10
                case 'V': n += 3 if p == 'I' else 5
                case 'I': n += 1
            p = c
        return n

    # def romanToInt(self, s: str) -> int:
    #     n = 0
    #     for c in s[::-1]:
    #         match c:
    #             case 'I': n += -1 if n >= 5 else 1
    #             case 'V': n += 5
    #             case 'X': n += -10 if n >= 50 else 10
    #             case 'L': n += 50
    #             case 'C': n += -100 if n >= 500 else 100
    #             case 'D': n += 500
    #             case 'M': n += 1000
    #     return n


if __name__ == '__main__':
    import unittest

    CASES = (
        ('III', 3),
        ('LVIII', 58),
        ('MCMXCIV', 1994),
    )

    class Test(unittest.TestCase):
        def test(self):
            romanToInt = Solution().romanToInt
            for s, excepted in CASES:
                actual = romanToInt(s)
                self.assertEqual(actual, excepted)

    unittest.main()

    from timeit import timeit

    romanToInt = Solution().romanToInt
    for s, _ in CASES:
        print(timeit(lambda: romanToInt, number=1000000))
